From db62c178eb51b8f29d02a9d540c85b3dfc2c998a Mon Sep 17 00:00:00 2001
From: Tor Vic <torvic9@mailbox.org>
Date: Mon, 7 Feb 2022 13:59:21 +0100
Subject: [PATCH 1/5] Kbuild, zstd: use a higher compression level for module
 compression

By default, zstd uses a compression level of 3.
Increase this level to 7 for a slightly better compression ratio, without
affecting compression speed too much.
---
 scripts/Makefile.modinst | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/scripts/Makefile.modinst b/scripts/Makefile.modinst
index ab0c5bd1a..1090e7d17 100644
--- a/scripts/Makefile.modinst
+++ b/scripts/Makefile.modinst
@@ -101,7 +101,7 @@ quiet_cmd_gzip = GZIP    $@
 quiet_cmd_xz = XZ      $@
       cmd_xz = $(XZ) --lzma2=dict=2MiB -f $<
 quiet_cmd_zstd = ZSTD    $@
-      cmd_zstd = $(ZSTD) -T0 --rm -f -q $<
+      cmd_zstd = $(ZSTD) -T0 -7 --rm -f -q $<
 
 $(dst)/%.ko.gz: $(dst)/%.ko FORCE
 	$(call cmd,gzip)
-- 
2.40.0.71.g950264636c


From 4bff4f948736a0584aedd3d143fed34914e3ad29 Mon Sep 17 00:00:00 2001
From: Tor Vic <torvic9@mailbox.org>
Date: Mon, 7 Feb 2022 13:54:13 +0100
Subject: [PATCH 2/5] Kbuild, zstd: allow users to choose a compression level
 for the kernel image

The zstd compression algorithm has a wide range of different compression
levels.
Give users the possibility to choose a custom compression level.
Also enable multithreading which is already enabled for module compression.
---
 init/Kconfig         | 13 +++++++++++++
 scripts/Makefile.lib | 13 ++++++++++---
 2 files changed, 23 insertions(+), 3 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index c88bb30a8..91269a9dd 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -348,6 +348,19 @@ config KERNEL_UNCOMPRESSED
 
 endchoice
 
+menu "ZSTD compression options"
+	depends on KERNEL_ZSTD
+
+config ZSTD_COMP_VAL
+	int "Compression level (1-22)"
+	range 1 22
+	default "22"
+	help
+	  Choose a compression level for zstd kernel compression.
+	  Default is 22, which is the maximum.
+
+endmenu
+
 config DEFAULT_INIT
 	string "Default init path"
 	default ""
diff --git a/scripts/Makefile.lib b/scripts/Makefile.lib
index 100a386fc..f34cb523b 100644
--- a/scripts/Makefile.lib
+++ b/scripts/Makefile.lib
@@ -542,14 +542,21 @@ quiet_cmd_xzmisc = XZMISC  $@
 # decompression is used, like initramfs decompression, zstd22 should likely not
 # be used because it would require zstd to allocate a 128 MB buffer.
 
+ifdef CONFIG_ZSTD_COMP_VAL
+zstd_comp_val	:= $(CONFIG_ZSTD_COMP_VAL)
+ifeq ($(shell test $(zstd_comp_val) -gt 19; echo $$?),0)
+zstd_comp_val	+= --ultra
+endif
+endif
+
 quiet_cmd_zstd = ZSTD    $@
-      cmd_zstd = cat $(real-prereqs) | $(ZSTD) -19 > $@
+      cmd_zstd = cat $(real-prereqs) | $(ZSTD) -T0 -19 > $@
 
 quiet_cmd_zstd22 = ZSTD22  $@
-      cmd_zstd22 = cat $(real-prereqs) | $(ZSTD) -22 --ultra > $@
+      cmd_zstd22 = cat $(real-prereqs) | $(ZSTD) -T0 -22 --ultra > $@
 
 quiet_cmd_zstd22_with_size = ZSTD22  $@
-      cmd_zstd22_with_size = { cat $(real-prereqs) | $(ZSTD) -22 --ultra; $(size_append); } > $@
+      cmd_zstd22_with_size = { cat $(real-prereqs) | $(ZSTD) -T0 -$(zstd_comp_val); $(size_append); } > $@
 
 # ASM offsets
 # ---------------------------------------------------------------------------
-- 
2.40.0.71.g950264636c


From 84811bc15353e48a6372b9cedac0a02dbc893d4f Mon Sep 17 00:00:00 2001
From: Tor Vic <torvic9@mailbox.org>
Date: Mon, 20 Feb 2023 11:06:09 +0100
Subject: [PATCH 3/5] Kbuild, zstd: Allow users to choose a compression level
 for modules

---
 kernel/module/Kconfig    | 8 ++++++++
 scripts/Makefile.modinst | 2 +-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/kernel/module/Kconfig b/kernel/module/Kconfig
index 424b3bc58..310cf5c1d 100644
--- a/kernel/module/Kconfig
+++ b/kernel/module/Kconfig
@@ -219,6 +219,14 @@ config MODULE_COMPRESS_ZSTD
 
 endchoice
 
+config MODULE_COMPRESS_ZSTD_LVL
+	int "Compression level (1-19)"
+	depends on MODULE_COMPRESS_ZSTD
+	range 1 19
+	default 7
+	help
+	  Compression level used by zstd for compressing modules.
+
 config MODULE_DECOMPRESS
 	bool "Support in-kernel module decompression"
 	depends on MODULE_COMPRESS_GZIP || MODULE_COMPRESS_XZ || MODULE_COMPRESS_ZSTD
diff --git a/scripts/Makefile.modinst b/scripts/Makefile.modinst
index 1090e7d17..102670905 100644
--- a/scripts/Makefile.modinst
+++ b/scripts/Makefile.modinst
@@ -101,7 +101,7 @@ quiet_cmd_gzip = GZIP    $@
 quiet_cmd_xz = XZ      $@
       cmd_xz = $(XZ) --lzma2=dict=2MiB -f $<
 quiet_cmd_zstd = ZSTD    $@
-      cmd_zstd = $(ZSTD) -T0 -7 --rm -f -q $<
+      cmd_zstd = $(ZSTD) -T0 -$(CONFIG_MODULE_COMPRESS_ZSTD_LVL) --rm -f -q $<
 
 $(dst)/%.ko.gz: $(dst)/%.ko FORCE
 	$(call cmd,gzip)
-- 
2.40.0.71.g950264636c


From 5cb170b48e582573a788192f5e923031ea8adadd Mon Sep 17 00:00:00 2001
From: Tor Vic <torvic9@mailbox.org>
Date: Mon, 20 Feb 2023 11:31:40 +0100
Subject: [PATCH 4/5] Kconfig, zstd: Rename COMP_VAL to COMPRESSION_LVL

---
 init/Kconfig         | 2 +-
 scripts/Makefile.lib | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index 91269a9dd..2f81d3827 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -351,7 +351,7 @@ endchoice
 menu "ZSTD compression options"
 	depends on KERNEL_ZSTD
 
-config ZSTD_COMP_VAL
+config ZSTD_COMPRESSION_LVL
 	int "Compression level (1-22)"
 	range 1 22
 	default "22"
diff --git a/scripts/Makefile.lib b/scripts/Makefile.lib
index f34cb523b..c13a218c5 100644
--- a/scripts/Makefile.lib
+++ b/scripts/Makefile.lib
@@ -542,8 +542,8 @@ quiet_cmd_xzmisc = XZMISC  $@
 # decompression is used, like initramfs decompression, zstd22 should likely not
 # be used because it would require zstd to allocate a 128 MB buffer.
 
-ifdef CONFIG_ZSTD_COMP_VAL
-zstd_comp_val	:= $(CONFIG_ZSTD_COMP_VAL)
+ifdef CONFIG_ZSTD_COMPRESSION_LVL
+zstd_comp_val	:= $(CONFIG_ZSTD_COMPRESSION_LVL)
 ifeq ($(shell test $(zstd_comp_val) -gt 19; echo $$?),0)
 zstd_comp_val	+= --ultra
 endif
-- 
2.40.0.71.g950264636c


From e8034222d6054cb9d4f5bdc0aa49db956d06401d Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 27 Mar 2023 13:25:22 +0200
Subject: [PATCH 5/5] kbuild-modules-6.3: use CachyOS settings

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig             |  2 +-
 kernel/module/Kconfig    | 23 ++++++++++++++++++++---
 scripts/Makefile.lib     |  4 ++--
 scripts/Makefile.modinst |  7 ++++++-
 4 files changed, 29 insertions(+), 7 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index 2f81d3827..2cd7503dd 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -351,7 +351,7 @@ endchoice
 menu "ZSTD compression options"
 	depends on KERNEL_ZSTD
 
-config ZSTD_COMPRESSION_LVL
+config ZSTD_COMPRESSION_LEVEL
 	int "Compression level (1-22)"
 	range 1 22
 	default "22"
diff --git a/kernel/module/Kconfig b/kernel/module/Kconfig
index 310cf5c1d..ecf2798c5 100644
--- a/kernel/module/Kconfig
+++ b/kernel/module/Kconfig
@@ -219,14 +219,31 @@ config MODULE_COMPRESS_ZSTD
 
 endchoice
 
-config MODULE_COMPRESS_ZSTD_LVL
+menu "ZSTD module compression options"
+    depends on MODULE_COMPRESS_ZSTD
+
+config MODULE_COMPRESS_ZSTD_LEVEL
 	int "Compression level (1-19)"
-	depends on MODULE_COMPRESS_ZSTD
 	range 1 19
-	default 7
+	default 9
 	help
 	  Compression level used by zstd for compressing modules.
 
+config MODULE_COMPRESS_ZSTD_ULTRA
+	bool "Enable ZSTD ultra compression"
+	help
+	  Compress modules with ZSTD using the highest possible compression.
+
+config MODULE_COMPRESS_ZSTD_LEVEL_ULTRA
+	int "Compression level (20-22)"
+	depends on MODULE_COMPRESS_ZSTD_ULTRA
+	range 20 22
+	default 20
+	help
+	  Ultra compression level used by zstd for compressing modules.
+
+endmenu
+
 config MODULE_DECOMPRESS
 	bool "Support in-kernel module decompression"
 	depends on MODULE_COMPRESS_GZIP || MODULE_COMPRESS_XZ || MODULE_COMPRESS_ZSTD
diff --git a/scripts/Makefile.lib b/scripts/Makefile.lib
index c13a218c5..a3ec7265f 100644
--- a/scripts/Makefile.lib
+++ b/scripts/Makefile.lib
@@ -542,8 +542,8 @@ quiet_cmd_xzmisc = XZMISC  $@
 # decompression is used, like initramfs decompression, zstd22 should likely not
 # be used because it would require zstd to allocate a 128 MB buffer.
 
-ifdef CONFIG_ZSTD_COMPRESSION_LVL
-zstd_comp_val	:= $(CONFIG_ZSTD_COMPRESSION_LVL)
+ifdef CONFIG_ZSTD_COMPRESSION_LEVEL
+zstd_comp_val	:= $(CONFIG_ZSTD_COMPRESSION_LEVEL)
 ifeq ($(shell test $(zstd_comp_val) -gt 19; echo $$?),0)
 zstd_comp_val	+= --ultra
 endif
diff --git a/scripts/Makefile.modinst b/scripts/Makefile.modinst
index 102670905..f4989f706 100644
--- a/scripts/Makefile.modinst
+++ b/scripts/Makefile.modinst
@@ -100,8 +100,13 @@ quiet_cmd_gzip = GZIP    $@
       cmd_gzip = $(KGZIP) -n -f $<
 quiet_cmd_xz = XZ      $@
       cmd_xz = $(XZ) --lzma2=dict=2MiB -f $<
+ifdef CONFIG_MODULE_COMPRESS_ZSTD_ULTRA
 quiet_cmd_zstd = ZSTD    $@
-      cmd_zstd = $(ZSTD) -T0 -$(CONFIG_MODULE_COMPRESS_ZSTD_LVL) --rm -f -q $<
+      cmd_zstd = $(ZSTD) -$(CONFIG_MODULE_COMPRESS_ZSTD_LEVEL_ULTRA) --ultra --zstd=wlog=21 -T0 --rm -f -q $<
+else
+quiet_cmd_zstd = ZSTD    $@
+      cmd_zstd = $(ZSTD) -$(CONFIG_MODULE_COMPRESS_ZSTD_LEVEL) --zstd=wlog=21 -T0 --rm -f -q $<
+endif
 
 $(dst)/%.ko.gz: $(dst)/%.ko FORCE
 	$(call cmd,gzip)
-- 
2.40.0.71.g950264636c

